### PR TITLE
Enforces deployment concurrency with orchestration polices.

### PR DESCRIPTION
**Pull request by jeanluciano**
Closes https://github.com/PrefectHQ/prefect/issues/15475

We want to expand the potential clients that can use deployment concurrency limits to 2.x, not just 3.x, and without requiring any code changes from the user, including package upgrades. To do this, we're transitioning from worker-initiated deployment concurrency handling to orchestration rules.

This code is based on `SecureTaskConcurrencySlots` and `ReleaseTaskConcurrencySlots` as reference implementations. If you, the reader, are unfamiliar with orchestration rules, this code is probably a little gnarly to understand. However, it's not radically different than what we did with workers before:

Unlike task run concurrency, flow run concurrency hooks into the transition from Scheduled to Pending (rather than Pending to Running). This is so that workers can avoid spinning up infrastructure for a flow run if they weren't able to secure a concurrency slot.
When clients propose a Pending state and we accept it, we'll check if the flow run is from a deployment, and if so, we'll try to take a concurrency slot.

If we can't take it, we'll either "enqueue" the blocked run or cancel it, based on the deployment's `concurrency_options.collision_strategy`. ENQUEUE moves the run to a `Scheduled` state named `AwaitingConcurrencySlot`, scheduled at a time in the future configurable with a setting (defaulting to "now"). ?`CANCEL_NEW` cancels the blocked run with a message about the concurrency limit.
If the limit is 0, that would deadlock, so we block reject the transition. Users can't submit a zero value through the deployment concurrency interface, but they can for global concurrency limits generally, so we protect against a weird case where a GCL with a zero limits gets attached to a deployment.

Releasing slots is a `BaseUniversalTransform` so that releasing slots happens even if another rule invalidates the transition. This can happen if, for example, flow retries are configured. The retries rule will mutate the transition to become Running -> Scheduled instead of the originally-proposed Running -> Failed. After a rule has mutated the transition, it becomes invalid when other rules see it, so they "fizzle." Fizzled rules run their cleanup method instead of their transition hook. However, BaseUniversalTransforms always run, even if the transition becomes invalid, so this is a way to write the release logic once (instead of in both a transition hook and cleanup method).

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

**Review by desertaxle — APPROVED** [review]
One small nit for an additional test assertion, but otherwise LGTM! Thanks for the thorough tests!

File: src/prefect/settings.py
Comment (zzstoatzz) [inline src/prefect/settings.py:1461]: why is the default 0?

**Review by zzstoatzz — COMMENTED** [review]
i more or less land in the same place here as @zangell44 where it generally looks good to me, but find it hard to think about

> What happens when a running flow run is deleted?
> What happens when a concurrency limit is set and there are already runs executing in progress?

File: src/prefect/settings.py
Comment (jeanluciano) [reply src/prefect/settings.py:1461]: We felt like latency was enough of a wait to avoid looping thru the same rescheduled run while at the same time not have unnecessary waiting to by runs that can't grab a slot.

**Review by cicdw — CHANGES_REQUESTED** [review]
request for a print to move to a log. 

Just want to call out that this will exacerbate the deadlocking problem when used with workers running the last few releases due to the duplicate slot acquisition logic across workers and runners.

**Comment by jeanluciano** [comment]
Changes to remove the worker and runner concurrency enforcement have been merged.

**Comment by abrookins** [comment]
In the Cloud PR, I increased the default retry period to wait 30 seconds.

**Comment by abrookins** [comment]
> i more or less land in the same place here as @zangell44 where it generally looks good to me, but find it hard to think about
> 
> > What happens when a running flow run is deleted?
> > What happens when a concurrency limit is set and there are already runs executing in progress?

1. Flow run deletion cleanup logic needs to include decrementing the slots if someone deletes a flow run in a pending, running, or cancelling state. I'm adding that to Cloud now and can backport though I don't think we should hold up the release for that. Your call!
 2. If I understood your question (and if I understand these two systems 😂 😅), the same things that would happen, more or less, when you change/add/remove task run concurrency limits with running tasks.

File: src/prefect/settings.py
Comment (abrookins) [reply src/prefect/settings.py]: Let's increase to 30 to match the Cloud PR.

File: src/prefect/server/orchestration/core_policy.py
Comment (abrookins) [inline src/prefect/server/orchestration/core_policy.py:409]: This needs to including PENDING as well.

**Review by abrookins — CHANGES_REQUESTED** [review]

**Comment by abrookins** [comment]
> Just want to call out that this will exacerbate the deadlocking problem when used with workers running the last few releases due to the duplicate slot acquisition logic across workers and runners.

@cicdw Want us to include the client version checking that Cloud uses, so we exclude clients on those versions?

**Comment by abrookins** [comment]
@jeanluciano This PR also didn't actually wire up the new rules to the flow policy. I'll do that with a client check to prevent deadlocks.

**Review by abrookins — APPROVED** [review]

**Comment by abrookins** [comment]
Hold up, we found an issue with the Prefect YAML file.

**Comment by abrookins** [comment]
@cicdw There's probably a more elegant way to fix this, but ruamel can't serialize enums and I wanted to get it working for now rather than futz with it. [`c93d85b` (#15504)](https://github.com/PrefectHQ/prefect/pull/15504/commits/c93d85b5d13cf3bcbcb8aff15e2533323eb7c4b7)

File: src/prefect/server/models/flow_runs.py
Comment (cicdw) [inline src/prefect/server/models/flow_runs.py:323]: the signature of this utility seems random to me - a `(deployment_id , final_state)` pairing is an ambiguous piece of state; presumably this would be better captured by a `flow_run` object or something?

**Review by cicdw — COMMENTED** [review]
one minor comment about the signature of an internal utility, otherwise i think this looks good

File: src/prefect/server/models/flow_runs.py
Comment (abrookins) [reply src/prefect/server/models/flow_runs.py:323]: Agreed. The flow run will have its deployment ID and final state.

**Review by cicdw — APPROVED** [review]

**Pull request merged**
